lib/Sema/CSSimplify.cpp

```diff
@@ -1579,25 +1579,32 @@ ConstraintSystem::matchSuperclassTypes(Type type1, Type type2,
   return getTypeMatchFailure(locator);
 }
 
-static ConstraintSystem::TypeMatchResult
-matchDeepTypeArguments(ConstraintSystem &cs,
-                       ConstraintSystem::TypeMatchOptions subflags,
-                       ArrayRef<Type> args1,
-                       ArrayRef<Type> args2,
-                       ConstraintLocatorBuilder locator) {
+static ConstraintSystem::TypeMatchResult matchDeepTypeArguments(
+    ConstraintSystem &cs, ConstraintSystem::TypeMatchOptions subflags,
+    ArrayRef<Type> args1, ArrayRef<Type> args2,
+    ConstraintLocatorBuilder locator,
+    llvm::function_ref<void(unsigned)> recordMismatch = [](unsigned) {}) {
   if (args1.size() != args2.size()) {
     return cs.getTypeMatchFailure(locator);
   }
+
+  ConstraintSystem::TypeMatchResult fullMatchResult = cs.getTypeMatchSuccess();
+
   for (unsigned i = 0, n = args1.size(); i != n; ++i) {
-    auto result = cs.matchTypes(args1[i], args2[i], ConstraintKind::Bind,
-                                subflags, locator.withPathElement(
-                                        LocatorPathElt::getGenericArgument(i)));
+    auto result = cs.matchTypes(
+        args1[i], args2[i], ConstraintKind::Bind, subflags,
+        locator.withPathElement(LocatorPathElt::getGenericArgument(i)));
 
-    if (result.isFailure())
-      return result;
+    if (result.isFailure()) {
+      if (!cs.shouldAttemptFixes())
+        return result;
+
+      fullMatchResult = result;
+      recordMismatch(i);
+    }
   }
 
-  return cs.getTypeMatchSuccess();
+  return fullMatchResult;
```

Comment:
This doesn't look right - `fullMatchResult` would be set to a failure if there was at least one we have recorded, but if it was "fixed" then result has to be access, I think there is no need to even have it here, we can just return success like before since argument matcher continues after failures when fixes are enabled.